File: bench.php
<?php
# bench.php -- the PHP entry in the language speed comparison.
#
# Usage: php bench.php <benchmark> <size> [reps]
# Prints: OK <benchmark> <checksum> <compute_milliseconds>
#
# Same algorithm, same deterministic input, same checksum as every other
# bench.* in this suite.
#
# Plain PHP 7.3+, no extensions required -- though the runner switches on
# opcache's tracing JIT when it is present. Two judgement calls worth
# knowing about: PHP integers are signed 64-bit and *overflow to float*
# instead of wrapping, so the shared PRNG's 64-bit multiply is done in
# 32-bit halves whose products stay safely inside 63 bits; and the sieve's
# flat byte array is a string poked one byte at a time, because a PHP array
# of 50 million ints costs ~16 bytes per entry where C spends 1 -- the
# mutable string is PHP's honest equivalent of a byte array, the way vec()
# is Perl's and typed arrays are JavaScript's.

# The CLI SAPI documents memory_limit as -1, but distro php.ini files (Nix,
# Debian) often cap it at 128M, which kills the sieve and quicksort at
# --heavy scale before they can print. Restore the documented CLI default.
ini_set('memory_limit', '-1');

# ---------- shared deterministic PRNG (identical in every language here) ----------
# State is one u64 held as two 32-bit halves. The multiplier splits as
# 0x5851F42D_4C957F2D and the increment as 0x14057B7E_F767814F; every partial
# product below is < 2^63, so nothing ever silently becomes a float.
$rng_lo = 0;
$rng_hi = 0;

function rng_seed($s) {
    global $rng_lo, $rng_hi;
    $rng_lo = $s & 0xFFFFFFFF;
    $rng_hi = ($s >> 32) & 0xFFFFFFFF;
}

function rng_next() {
    global $rng_lo, $rng_hi;
    $p  = $rng_lo * 0x4C957F2D;
    $lo = ($p & 0xFFFFFFFF) + 0xF767814F;
    $hi = ($p >> 32)
        + (($rng_lo * 0x5851F42D) & 0xFFFFFFFF)
        + (($rng_hi * 0x4C957F2D) & 0xFFFFFFFF)
        + 0x14057B7E + ($lo >> 32);
    $rng_lo = $lo & 0xFFFFFFFF;
    $rng_hi = $hi & 0xFFFFFFFF;
    return $rng_hi >> 1;    # top 31 bits of the 64-bit state
}

# ---------- 1. mandelbrot: tight floating-point loop, zero allocation ----------
function bench_mandelbrot($n) {
    $total = 0;
    for ($py = 0; $py < $n; $py++) {
        $ci = 2.0 * $py / $n - 1.0;
        for ($px = 0; $px < $n; $px++) {
            $cr = 2.0 * $px / $n - 1.5;
            $zr = 0.0;
            $zi = 0.0;
            $i = 0;
            while ($i < 255) {
                $zr2 = $zr * $zr;
                $zi2 = $zi * $zi;
                if ($zr2 + $zi2 > 4.0) break;
                $zi = 2.0 * $zr * $zi + $ci;
                $zr = $zr2 - $zi2 + $cr;
                $i++;
            }
            $total += $i;
        }
    }
    return $total;
}

# ---------- 2. sieve of eratosthenes: integer math over a large flat array ----------
function bench_sieve($n) {
    $comp = str_repeat("\0", $n + 1);    # one byte per candidate, like everyone else
    $count = 0;
    for ($i = 2; $i <= $n; $i++) {
        if ($comp[$i] === "\0") {
            $count++;
            for ($j = $i * $i; $j <= $n; $j += $i) {
                $comp[$j] = "\1";
            }
        }
    }
    return $count;
}

# ---------- 3. quicksort: branches, swaps, recursion, random memory access ----------
# Hand-written, not sort() -- the point is to measure the language, not the
# quality of its sort library.
function insertion_sort(&$a, $lo, $hi) {
    for ($i = $lo + 1; $i <= $hi; $i++) {
        $v = $a[$i];
        $j = $i - 1;
        while ($j >= $lo && $a[$j] > $v) {
            $a[$j + 1] = $a[$j];
            $j--;
        }
        $a[$j + 1] = $v;
    }
}

function quicksort(&$a, $lo, $hi) {
    while ($hi - $lo > 16) {
        $mid = $lo + (($hi - $lo) >> 1);
        # median-of-three: order a[lo] <= a[mid] <= a[hi]
        if ($a[$mid] < $a[$lo]) { $t = $a[$mid]; $a[$mid] = $a[$lo];  $a[$lo]  = $t; }
        if ($a[$hi]  < $a[$lo]) { $t = $a[$hi];  $a[$hi]  = $a[$lo];  $a[$lo]  = $t; }
        if ($a[$hi] < $a[$mid]) { $t = $a[$hi];  $a[$hi]  = $a[$mid]; $a[$mid] = $t; }
        $pivot = $a[$mid];

        $i = $lo;
        $j = $hi;
        while ($i <= $j) {
            while ($a[$i] < $pivot) $i++;
            while ($a[$j] > $pivot) $j--;
            if ($i <= $j) {
                $t = $a[$i]; $a[$i] = $a[$j]; $a[$j] = $t;
                $i++;
                $j--;
            }
        }
        # recurse into the smaller half, loop on the larger: depth stays O(log n)
        if ($j - $lo < $hi - $i) { quicksort($a, $lo, $j); $lo = $i; }
        else                     { quicksort($a, $i, $hi); $hi = $j; }
    }
    insertion_sort($a, $lo, $hi);
}

function bench_quicksort($n) {
    global $rng_lo, $rng_hi;
    rng_seed(12345);
    # The PRNG body is pasted inline with the state in plain locals: `global`
    # binds by reference, and every read through a reference zval blunts the
    # JIT's type inference. Same expansion in wordcount and matmul.
    $slo = $rng_lo;
    $shi = $rng_hi;
    $a = array_fill(0, $n, 0);    # preallocated: no rehash mid-fill
    for ($i = 0; $i < $n; $i++) {
        $p   = $slo * 0x4C957F2D;
        $t   = ($p & 0xFFFFFFFF) + 0xF767814F;
        $shi = (($p >> 32)
             + (($slo * 0x5851F42D) & 0xFFFFFFFF)
             + (($shi * 0x4C957F2D) & 0xFFFFFFFF)
             + 0x14057B7E + ($t >> 32)) & 0xFFFFFFFF;
        $slo = $t & 0xFFFFFFFF;
        $a[$i] = $shi >> 1;
    }
    $rng_lo = $slo;
    $rng_hi = $shi;
    quicksort($a, 0, $n - 1);
    $h = 0;
    foreach ($a as $v) {
        $h = ($h * 31 + $v) & 0xFFFFFFFF;    # order-sensitive checksum
    }
    return $h;
}

# ---------- 4. word count: string hashing + hash map ----------
# The associative array is PHP's native data structure and it is written in
# C, so this is the benchmark where PHP gives up the least.
const VOCAB = 5000;

function bench_wordcount($n) {
    rng_seed(12345);
    $words = [];
    for ($i = 0; $i < VOCAB; $i++) {
        $len = 3 + rng_next() % 6;
        $w = '';
        for ($c = 0; $c < $len; $c++) $w .= chr(97 + rng_next() % 26);
        $words[] = $w;
    }

    global $rng_lo, $rng_hi;
    $slo = $rng_lo;    # PRNG inlined; see bench_quicksort
    $shi = $rng_hi;
    $counts = [];
    $maxc = 0;
    for ($k = 0; $k < $n; $k++) {
        $p   = $slo * 0x4C957F2D;
        $t   = ($p & 0xFFFFFFFF) + 0xF767814F;
        $shi = (($p >> 32)
             + (($slo * 0x5851F42D) & 0xFFFFFFFF)
             + (($shi * 0x4C957F2D) & 0xFFFFFFFF)
             + 0x14057B7E + ($t >> 32)) & 0xFFFFFFFF;
        $slo = $t & 0xFFFFFFFF;
        $ra = ($shi >> 1) % VOCAB;
        $p   = $slo * 0x4C957F2D;
        $t   = ($p & 0xFFFFFFFF) + 0xF767814F;
        $shi = (($p >> 32)
             + (($slo * 0x5851F42D) & 0xFFFFFFFF)
             + (($shi * 0x4C957F2D) & 0xFFFFFFFF)
             + 0x14057B7E + ($t >> 32)) & 0xFFFFFFFF;
        $slo = $t & 0xFFFFFFFF;
        $rb = ($shi >> 1) % VOCAB;
        # (int) of the float division equals intdiv here -- $ra*$rb < 2.5e7,
        # doubles are exact that low -- and it skips a real function call
        $w = $words[(int)($ra * $rb / VOCAB)];    # triangular, so counts vary
        $c = ($counts[$w] ?? 0) + 1;
        $counts[$w] = $c;
        if ($c > $maxc) $maxc = $c;
    }
    $rng_lo = $slo;
    $rng_hi = $shi;
    return count($counts) * 1000003 + $maxc;
}

# ---------- 5. binary trees: allocation, pointer chasing, memory management ----------
# Nothing here computes; the entire cost is allocating nodes and letting the
# refcounter take them back. Like Perl, PHP frees by reference counting, so
# the teardown cost lands right here in the loop when $t is reassigned.
class Node {
    public $l;
    public $r;
    function __construct($l, $r) { $this->l = $l; $this->r = $r; }
}

function make_tree($d) {
    if ($d === 0) return new Node(null, null);
    return new Node(make_tree($d - 1), make_tree($d - 1));
}

function check_tree($t) {
    if ($t->l === null) return 1;
    return 1 + check_tree($t->l) + check_tree($t->r);
}

function bench_binarytrees($n) {
    $h = 0;
    for ($k = 0; $k < $n; $k++) {
        $t = make_tree(11);    # 4,095 nodes, built and thrown away
        $h = ($h * 31 + check_tree($t) + $k) & 0xFFFFFFFF;
    }
    return $h;
}

# ---------- 6. matmul: triple-nested loops over flat 2D arrays ----------
function bench_matmul($n) {
    global $rng_lo, $rng_hi;
    rng_seed(12345);
    $nn = $n * $n;
    $slo = $rng_lo;    # PRNG inlined; see bench_quicksort
    $shi = $rng_hi;
    $a = array_fill(0, $nn, 0);
    $b = array_fill(0, $nn, 0);
    for ($i = 0; $i < $nn; $i++) {
        $p   = $slo * 0x4C957F2D;
        $t   = ($p & 0xFFFFFFFF) + 0xF767814F;
        $shi = (($p >> 32)
             + (($slo * 0x5851F42D) & 0xFFFFFFFF)
             + (($shi * 0x4C957F2D) & 0xFFFFFFFF)
             + 0x14057B7E + ($t >> 32)) & 0xFFFFFFFF;
        $slo = $t & 0xFFFFFFFF;
        $a[$i] = ($shi >> 1) % 100;
    }
    for ($i = 0; $i < $nn; $i++) {
        $p   = $slo * 0x4C957F2D;
        $t   = ($p & 0xFFFFFFFF) + 0xF767814F;
        $shi = (($p >> 32)
             + (($slo * 0x5851F42D) & 0xFFFFFFFF)
             + (($shi * 0x4C957F2D) & 0xFFFFFFFF)
             + 0x14057B7E + ($t >> 32)) & 0xFFFFFFFF;
        $slo = $t & 0xFFFFFFFF;
        $b[$i] = ($shi >> 1) % 100;
    }
    $rng_lo = $slo;
    $rng_hi = $shi;
    $c = array_fill(0, $nn, 0);
    for ($i = 0; $i < $n; $i++) {
        $ib = $i * $n;
        for ($j = 0; $j < $n; $j++) {
            $s = 0;
            $bi = $j;    # walks b down the column; saves a multiply per step
            for ($k = 0; $k < $n; $k++) {
                $s += $a[$ib + $k] * $b[$bi];
                $bi += $n;
            }
            $c[$ib + $j] = $s;
        }
    }
    $h = 0;
    for ($i = 0; $i < $nn; $i++) {
        $h = ($h * 31 + $c[$i]) & 0xFFFFFFFF;
    }
    return $h;
}

# ---------- main ----------
const BENCHMARKS = [
    'mandelbrot'  => 'bench_mandelbrot',
    'sieve'       => 'bench_sieve',
    'quicksort'   => 'bench_quicksort',
    'wordcount'   => 'bench_wordcount',
    'binarytrees' => 'bench_binarytrees',
    'matmul'      => 'bench_matmul',
];

function main($argv) {
    if (count($argv) < 3) {
        fwrite(STDERR, "usage: php bench.php <benchmark> <size> [reps]\n");
        return 2;
    }
    $name = $argv[1];
    $size = (int)$argv[2];
    $reps = isset($argv[3]) ? (int)$argv[3] : 1;
    if ($reps < 1) $reps = 1;

    if (!isset(BENCHMARKS[$name])) {
        fwrite(STDERR, "unknown benchmark: $name\n");
        return 2;
    }
    $fn = BENCHMARKS[$name];

    # Best-of-N: the fastest run is the one least polluted by scheduler noise,
    # cold caches and (for the JIT languages) not-yet-compiled code.
    $best = INF;
    $result = null;
    for ($r = 0; $r < $reps; $r++) {
        $t0 = hrtime(true);
        $v = $fn($size);
        $elapsed = (hrtime(true) - $t0) / 1e6;
        if ($elapsed < $best) $best = $elapsed;
        if ($r === 0) {
            $result = $v;
        } elseif ($v !== $result) {
            fwrite(STDERR, "nondeterministic result!\n");
            return 3;
        }
    }

    printf("OK %s %d %.3f\n", $name, $result, $best);
    return 0;
}

exit(main($argv));

File: bench.swift
/*
 * bench.swift -- the Swift entry in the language speed comparison.
 *
 * Usage: bench <benchmark> <size> [reps]
 * Prints: OK <benchmark> <checksum> <best_ms> <median_ms> <worst_ms>
 *
 * Same algorithm, same deterministic input, same checksum as every other
 * bench.* in this suite.
 *
 * Plain Swift, no Foundation, no packages: compiled with a bare `swiftc -O`.
 * Arrays keep their bounds checks (like Rust); the wrapping operators
 * (&*, &+) are the language's own way of asking for C's overflow semantics,
 * not an unsafe door.
 */

#if canImport(Glibc)
import Glibc
#elseif canImport(Darwin)
import Darwin
#endif

func nowMS() -> Double {
    var ts = timespec()
    clock_gettime(CLOCK_MONOTONIC, &ts)
    return Double(ts.tv_sec) * 1000.0 + Double(ts.tv_nsec) / 1e6
}

// ---------- shared deterministic PRNG (identical in every language here) ----------
var rngState: UInt64 = 0

func rngSeed(_ s: UInt64) { rngState = s }

func rngNext() -> UInt32 {
    rngState = rngState &* 6364136223846793005 &+ 1442695040888963407
    return UInt32(truncatingIfNeeded: rngState >> 33)  // top 31 bits
}

// ---------- 1. mandelbrot: tight floating-point loop, zero allocation ----------
func benchMandelbrot(_ n: Int) -> Int64 {
    var total: Int64 = 0
    for py in 0..<n {
        let ci = 2.0 * Double(py) / Double(n) - 1.0
        for px in 0..<n {
            let cr = 2.0 * Double(px) / Double(n) - 1.5
            var zr = 0.0, zi = 0.0
            var i = 0
            while i < 255 {
                let zr2 = zr * zr, zi2 = zi * zi
                if zr2 + zi2 > 4.0 { break }
                zi = 2.0 * zr * zi + ci
                zr = zr2 - zi2 + cr
                i += 1
            }
            total += Int64(i)
        }
    }
    return total
}

// ---------- 2. sieve of eratosthenes: integer math over a large flat array ----------
func benchSieve(_ n: Int) -> Int64 {
    var comp = [Bool](repeating: false, count: n + 1)
    var count: Int64 = 0
    var i = 2
    while i <= n {
        if !comp[i] {
            count += 1
            var j = i * i
            while j <= n {
                comp[j] = true
                j += i
            }
        }
        i += 1
    }
    return count
}

// ---------- 3. quicksort: branches, swaps, recursion, random memory access ----------
func insertionSort(_ a: inout [UInt32], _ lo: Int, _ hi: Int) {
    var i = lo + 1  // not a closed range: hi < lo happens on empty partitions
    while i <= hi {
        let v = a[i]
        var j = i - 1
        while j >= lo && a[j] > v {
            a[j + 1] = a[j]
            j -= 1
        }
        a[j + 1] = v
        i += 1
    }
}

func quicksort(_ a: inout [UInt32], _ lo: Int, _ hi: Int) {
    var lo = lo, hi = hi
    while hi - lo > 16 {
        let mid = lo + (hi - lo) / 2
        // median-of-three: order a[lo] <= a[mid] <= a[hi]
        if a[mid] < a[lo] { a.swapAt(mid, lo) }
        if a[hi] < a[lo] { a.swapAt(hi, lo) }
        if a[hi] < a[mid] { a.swapAt(hi, mid) }
        let pivot = a[mid]

        var i = lo, j = hi
        while i <= j {
            while a[i] < pivot { i += 1 }
            while a[j] > pivot { j -= 1 }
            if i <= j {
                a.swapAt(i, j)
                i += 1
                j -= 1
            }
        }
        // recurse into the smaller half, loop on the larger: depth stays O(log n)
        if j - lo < hi - i {
            quicksort(&a, lo, j)
            lo = i
        } else {
            quicksort(&a, i, hi)
            hi = j
        }
    }
    insertionSort(&a, lo, hi)
}

func benchQuicksort(_ n: Int) -> Int64 {
    rngSeed(12345)
    // filled by index: append() pays a uniqueness + capacity check per
    // element even after reserveCapacity, and this loop is inside the timing
    var a = [UInt32](repeating: 0, count: n)
    for i in 0..<n { a[i] = rngNext() }
    quicksort(&a, 0, n - 1)
    var h: UInt32 = 0
    for i in 0..<n { h = h &* 31 &+ a[i] }  // order-sensitive checksum
    return Int64(h)
}

// ---------- 4. word count: string hashing + hash map ----------
// Swift's Dictionary and String hashing live in the standard library
// (SipHash-1-3, like Rust's default), so this row prices that default.
let VOCAB = 5000

func benchWordcount(_ n: Int) -> Int64 {
    rngSeed(12345)
    var words = [String]()
    words.reserveCapacity(VOCAB)
    for _ in 0..<VOCAB {
        let len = 3 + Int(rngNext() % 6)
        // build as UTF-8 bytes: appending Characters one at a time drags in
        // grapheme breaking, which is pure overhead for plain ASCII
        var bytes = [UInt8]()
        bytes.reserveCapacity(len)
        for _ in 0..<len {
            bytes.append(UInt8(97 + rngNext() % 26))
        }
        words.append(String(decoding: bytes, as: UTF8.self))
    }

    var counts = [String: Int](minimumCapacity: 8192)
    var maxc = 0
    for _ in 0..<n {
        let ra = Int(rngNext()) % VOCAB
        let rb = Int(rngNext()) % VOCAB
        let w = words[(ra * rb) / VOCAB]  // triangular, so counts vary
        // one hash + probe via the index, not the two a read-then-write
        // subscript pair costs; values[i] updates in place, no rehash
        if let i = counts.index(forKey: w) {
            let c = counts.values[i] + 1
            counts.values[i] = c
            if c > maxc { maxc = c }
        } else {
            counts[w] = 1
            if maxc < 1 { maxc = 1 }
        }
    }
    return Int64(counts.count) * 1000003 + Int64(maxc)
}

// ---------- 5. binary trees: allocation, pointer chasing, memory management ----------
// Nothing here computes; the whole cost is class instances and ARC retain/
// release traffic -- Swift pays per-object like CPython's refcounts, not in
// GC bursts like Java.
final class Node {
    var l: Node?
    var r: Node?
}

func makeTree(_ d: Int) -> Node {
    let t = Node()
    if d > 0 {
        t.l = makeTree(d - 1)
        t.r = makeTree(d - 1)
    }
    return t
}

func checkTree(_ t: Node) -> Int {
    // test only l, like every other language: binding r too would retain/
    // release a second child per node, ARC traffic no one else pays
    guard let l = t.l else { return 1 }
    return 1 + checkTree(l) + checkTree(t.r!)
}

func benchBinarytrees(_ n: Int) -> Int64 {
    var h: UInt32 = 0
    for k in 0..<n {
        let t = makeTree(11)  // 4,095 nodes, built and thrown away
        h = h &* 31 &+ UInt32(checkTree(t)) &+ UInt32(k)
    }
    return Int64(h)
}

// ---------- 6. matmul: triple-nested loops over flat 2D arrays ----------
func benchMatmul(_ n: Int) -> Int64 {
    rngSeed(12345)
    var a = [UInt32](repeating: 0, count: n * n)
    var b = [UInt32](repeating: 0, count: n * n)
    var c = [UInt32](repeating: 0, count: n * n)
    for i in 0..<(n * n) { a[i] = rngNext() % 100 }
    for i in 0..<(n * n) { b[i] = rngNext() % 100 }
    for i in 0..<n {
        let ib = i * n
        for j in 0..<n {
            var s: UInt32 = 0
            var bi = j  // walks b down the column; saves a multiply per step
            for k in 0..<n { s = s &+ a[ib + k] &* b[bi]; bi += n }
            c[ib + j] = s
        }
    }
    var h: UInt32 = 0
    for i in 0..<(n * n) { h = h &* 31 &+ c[i] }
    return Int64(h)
}

// ---------- hidden: prng conformance check, not part of the scored suite ----------
// run.py --selftest uses it to validate each language's PRNG directly; several
// languages implement the 64-bit multiply in 32-bit halves and this checks every bit.
func benchPrng(_ n: Int) -> Int64 {
    rngSeed(12345)
    var h: UInt32 = 0
    for _ in 0..<n { h = h &* 31 &+ rngNext() }
    return Int64(h)
}

// ---------- driver ----------
func dispatch(_ name: String, _ size: Int) -> Int64? {
    switch name {
    case "mandelbrot": return benchMandelbrot(size)
    case "sieve": return benchSieve(size)
    case "quicksort": return benchQuicksort(size)
    case "wordcount": return benchWordcount(size)
    case "binarytrees": return benchBinarytrees(size)
    case "matmul": return benchMatmul(size)
    case "prng": return benchPrng(size)
    default: return nil
    }
}

func fmt3(_ x: Double) -> String {
    // "%.3f" without pulling in Foundation for String(format:)
    let t = Int((x * 1000.0).rounded())
    return "\(t / 1000)." + String(1000 + t % 1000).dropFirst()
}

func fail(_ msg: String, _ code: Int32) -> Never {
    fputs(msg + "\n", stderr)
    exit(code)
}

let args = CommandLine.arguments
if args.count < 3 {
    fail("usage: \(args[0]) <benchmark> <size> [reps]", 2)
}
let name = args[1]
guard let size = Int(args[2]) else { fail("bad size: \(args[2])", 2) }
let reps = max(args.count > 3 ? (Int(args[3]) ?? 1) : 1, 1)

// Best-of-N: the fastest run is the one least polluted by scheduler noise
// and cold caches. Median and worst ride along so the runner can report
// the spread.
var times = [Double]()
times.reserveCapacity(reps)
var result: Int64 = 0
for r in 0..<reps {
    let t0 = nowMS()
    guard let v = dispatch(name, size) else { fail("unknown benchmark: \(name)", 2) }
    times.append(nowMS() - t0)
    if r == 0 {
        result = v
    } else if v != result {
        fail("nondeterministic result!", 3)
    }
}

times.sort()
print("OK \(name) \(result) \(fmt3(times[0])) \(fmt3(times[reps / 2])) \(fmt3(times[reps - 1]))")
